### PR TITLE
Add ready check for Webhook server

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area open-source
/kind enhancement

**What this PR does / why we need it**:
The PR adds ready check for webhook server. The ready check wait for `10 sec` for webhook server to be ready, otherwise throws error.

**Which issue(s) this PR fixes**:
Part of #4772 

**Special notes for your reviewer**:
Since `StartedChecker()` is part of Kubernetes. I don't see any need to keep an extra function that just wraps around it.
**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Review by rfranzke — COMMENTED** [review]
/invite @timebertt 

@acumino Should we now also define the `readinessProbe`s in the respecitve `Deployment` specs?

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go [hunk 1]
```diff
@@ -174,6 +174,11 @@ func (o *options) run(ctx context.Context) error {
 	log.Info("setting up webhook server")
 	server := mgr.GetWebhookServer()
 
+	log.Info("setting up readycheck for webhook server")
+	if err := mgr.AddHealthzCheck("readyz", server.StartedChecker()); err != nil {
```
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go]: This should be added to the readiness endpoint not the health endpoint.
```suggestion
	if err := mgr.AddReadyzCheck("webhook-server", server.StartedChecker()); err != nil {
```

Health endpoint is for checking if the container is still alive (or might need to be restarted). The readiness endpoint is for determining whether the pod should be added to the ready set of containers to serve traffic for the service (and for other things like continuing a rolling update). This is what we want for the case of webhooks.

**Review by timebertt — CHANGES_REQUESTED** [review]
Thanks.

File: cmd/gardener-extension-provider-local/app/app.go [hunk 2]
```diff
@@ -233,6 +233,10 @@ func NewControllerManagerCommand(ctx context.Context) *cobra.Command {
 			reconcileOpts.Completed().Apply(&oscommon.DefaultAddOptions.IgnoreOperationAnnotation)
 			reconcileOpts.Completed().Apply(&localworker.DefaultAddOptions.IgnoreOperationAnnotation)
 
+			if err := mgr.AddReadyzCheck("webhook-server", mgr.GetWebhookServer().StartedChecker()); err != nil {
+				controllercmd.LogErrAndExit(err, "Could not add readycheck of webhook to manager")
```
Comment (rfranzke) [inline cmd/gardener-extension-provider-local/app/app.go]: FYI: With #5442 this line will look different.

**Comment by rfranzke** [comment]
@acumino Can you check why the e2e test is failing?

**Comment by acumino** [comment]
> @acumino Can you check why the e2e test is failing?

I am unable to figure it out from Prow logs. But probably the case I didn't add an endpoint in provider local. But I tried running it locally it is working fine. Any suggestions why?

**Comment by acumino** [comment]
@rfranzke If we want to add `ready check` in extension also, How to proceed? 

One way will be what I have done in `provide-local` i.e add `HealthProbeBindAddress` this way in each extension
```
var opts manager.Options = mgrOpts.Completed().Options()
opts.HealthProbeBindAddress = ":8081"
```
Or another way is to add this option [here](https://github.com/gardener/gardener/blob/24679a734ed55766d9c325da88a05274f3ae7f9f/extensions/pkg/controller/cmd/options.go#L153). Also do we want a flag for this to make it user-configurable or make it fixed to `8081`.

**Comment by rfranzke** [comment]
I think we should expose the port via a flag so that operators can change it if needed (no hard-coding). I think this is also still missing for `provider-local` IIUC. Other than this, you would have to touch all the extensions with webhook servers anyways to add the `readinessProbe` to their `deployment.yaml`, so you could do similar changes as for `provider-local` also in their respective Golang files.

File: charts/gardener/provider-local/extension/templates/deployment.yaml [hunk 3]
```diff
@@ -72,6 +72,16 @@ spec:
         - name: IMAGEVECTOR_OVERWRITE
           value: /charts_overwrite/images_overwrite.yaml
         {{- end }}
+        readinessProbe:
```
Comment (timebertt) [inline charts/gardener/provider-local/extension/templates/deployment.yaml:84]: There should be a `livenessProbe` here as well.

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 4]
```diff
@@ -252,6 +261,20 @@ func (g *gardenerSeedAdmissionController) Deploy(ctx context.Context) error {
 									corev1.ResourceMemory: resource.MustParse("100Mi"),
 								},
 							},
+							ReadinessProbe: &corev1.Probe{
```
Comment (timebertt) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:275]: Same here

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go [hunk 5]
```diff
@@ -557,6 +557,20 @@ func (r *resourceManager) ensureDeployment(ctx context.Context) error {
 							SuccessThreshold:    1,
 							TimeoutSeconds:      5,
 						},
+						ReadinessProbe: &corev1.Probe{
```
Comment (timebertt) [inline pkg/operation/botanist/component/resourcemanager/resource_manager.go:560]: Same here

**Review by timebertt — DISMISSED** [review]
Please ensure, that we have `readinessProbes` (webhook checks, informer readiness, etc.) and `livenessProbes` (basic ping check or others if fitting) for all components.

File: charts/gardener/provider-local/extension/templates/deployment.yaml (hunk 3 above)
Comment (acumino) [reply charts/gardener/provider-local/extension/templates/deployment.yaml:84]: Is it necessary to have `livenessProbe ` if there is `readinessProbe `?

Comment (timebertt) [reply charts/gardener/provider-local/extension/templates/deployment.yaml:84]: Yes. `livenessProbe` is for restarting containers if they are stuck. `readinessProbe` is for determining when to start serving traffic and continue rollouts. 
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/

Comment (acumino) [reply charts/gardener/provider-local/extension/templates/deployment.yaml:84]: for readiness probe:
Webhook server check
informer for cache
anything else?

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 6]
```diff
@@ -238,6 +247,7 @@ func (g *gardenerSeedAdmissionController) Deploy(ctx context.Context) error {
 								fmt.Sprintf("--port=%d", port),
 								fmt.Sprintf("--tls-cert-dir=%s", volumeMountPath),
 								fmt.Sprintf("--allow-invalid-extension-resources=%t", !gardenletfeatures.FeatureGate.Enabled(features.DenyInvalidExtensionResources)),
+								fmt.Sprintf("--health-bind-address=:%v", healthPort),
```
Comment (rfranzke) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]:
```suggestion
								fmt.Sprintf("--health-bind-address=:%d", healthPort),
```

File: cmd/gardener-extension-provider-local/app/app.go [hunk 7]
```diff
@@ -262,7 +278,9 @@ func NewControllerManagerCommand(ctx context.Context) *cobra.Command {
 		},
 	}
 
-	aggOption.AddFlags(cmd.Flags())
+	flags := cmd.Flags()
+	aggOption.AddFlags(flags)
+	flags.StringVar(&opts.HealthProbeBindAddress, "health-bind-address", ":8081", "bind address for the health server")
```
Comment (rfranzke) [inline cmd/gardener-extension-provider-local/app/app.go]: Shouldn't this flag go to the manager options?

Comment (acumino) [reply cmd/gardener-extension-provider-local/app/app.go]: Yes but then changes need to be made [here](https://github.com/gardener/gardener/blob/cf1439242213e86fe400c83c10e05943d759d72a/extensions/pkg/controller/cmd/options.go#L185). And everything that usages extension/pkg/cmd will have that flag. If we are good with it, I think here https://github.com/gardener/gardener/pull/5380#issuecomment-1041179214 we should choose the second option, i.e. add `HealthProbeBindAddress` to manager options.

Comment (rfranzke) [reply cmd/gardener-extension-provider-local/app/app.go]: Yeah, I think it makes sense that everybody gets this automatically after revendoring (less duplicating in `n` extension repos).

Comment (acumino) [reply cmd/gardener-extension-provider-local/app/app.go]: I will update the PR accordingly.

Comment (acumino) [reply cmd/gardener-extension-provider-local/app/app.go]: 3435b89f64ba2610bb08f6626c6200a027565a48 I have updated the changes.

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: charts/gardener/provider-local/extension/templates/deployment.yaml [hunk 8]
```diff
@@ -72,6 +75,26 @@ spec:
...
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: {{ .Values.healthPort }}
+            scheme: HTTP
+          initialDelaySeconds: 30
+          failureThreshold: 5
+          periodSeconds: 10
+          successThreshold: 1
+          timeoutSeconds: 5
```
Comment (timebertt) [inline charts/gardener/provider-local/extension/templates/deployment.yaml]: I suggest the following values (stick to [defaults](https://github.com/kubernetes/kubernetes/blob/release-1.23/pkg/apis/core/v1/defaults.go#L207-L220) for the rest):
```suggestion
          initialDelaySeconds: 10
```

File: charts/gardener/provider-local/extension/templates/deployment.yaml [hunk 9]
```diff
@@ -72,6 +75,26 @@ spec:
...
+        readinessProbe:
+          httpGet:
+            path: /readyz
+            port: {{ .Values.healthPort }}
+            scheme: HTTP
+          initialDelaySeconds: 5
+          failureThreshold: 5
+          periodSeconds: 5
+          successThreshold: 1
+          timeoutSeconds: 5
```
Comment (timebertt) [inline charts/gardener/provider-local/extension/templates/deployment.yaml]:
```suggestion
          initialDelaySeconds: 5
```

File: cmd/gardener-resource-manager/app/app.go [hunk 10]
```diff
@@ -118,6 +117,11 @@ func NewResourceManagerCommand() *cobra.Command {
 					return fmt.Errorf("could not add target cluster to manager: %w", err)
 				}
 
+				log.Info("Setting up readycheck for webhook server")
```
Comment (timebertt) [inline cmd/gardener-resource-manager/app/app.go:130]: I believe the ping check is missing for `/healthz` in grm.

Comment (acumino) [reply cmd/gardener-resource-manager/app/app.go:130]: I think it's not needed as we already add this [here](https://github.com/gardener/gardener/blob/744e0ff5db4f7f58d70a2f43546d379303f8a2cf/cmd/gardener-resource-manager/app/app.go#L132).

Comment (timebertt) [reply cmd/gardener-resource-manager/app/app.go:130]: I don't see, where the ping check is added in the link you provided

Comment (acumino) [reply cmd/gardener-resource-manager/app/app.go:130]: No, I said if this check is there, probably there is no need for ping.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**